### PR TITLE
Add BANFrance Geocoder

### PR DESCRIPTION
**Pull request by sheub**
Hi there,
A Geocoder service from the  adresse.data.gouv.fr 
It does not require any Api_Key nore registration.
cheers...
sheub

### CODE REVIEW

File: geopy/geocoders/banfrance.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+        :param str format_string:
+            See :attr:`geopy.geocoders.options.default_format_string`.
+
+            .. versionadded:: 1.14.0
+
+        :type ssl_context: :class:`ssl.SSLContext`
+        :param ssl_context:
+            See :attr:`geopy.geocoders.options.default_ssl_context`.
+
+            .. versionadded:: 1.14.0
```
Comment (KostyaEsmukov) [inline geopy/geocoders/banfrance.py]: Could you please remove these version-related directives? This class didn't even exist in these versions, so these arguments couldn't have been technically added to this class in these versions.

File: geopy/geocoders/banfrance.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+            '%s://%s%s' % (self.scheme, self.domain, self.geocode_path)
+        )
+        self.reverse_api = (
+            '%s://%s%s' % (self.scheme, self.domain, self.reverse_path)
+        )
+
+    def geocode(
+            self,
+            query,
+            limit=None,
```
Comment (KostyaEsmukov) [inline geopy/geocoders/banfrance.py]: This parameter seems to be ignored: I believe the implementation is accidentally missing.

Also the corresponding testcase is missing, could you please add one? You could use this one for inspiration: https://github.com/geopy/geopy/blob/3b19c8e4f6fdf13dddfbe7a3703ba031ec95194d/test/geocoders/nominatim.py#L48-L60

File: geopy/geocoders/banfrance.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+        logger.debug("%s.geocode: %s", self.__class__.__name__, url)
+        return self._parse_json(
+            self._call_geocoder(url, timeout=timeout), exactly_one
+        )
+
+    def reverse(
+            self,
+            query,
+            exactly_one=DEFAULT_SENTINEL,
```
Comment (KostyaEsmukov) [inline geopy/geocoders/banfrance.py]: It should be just `exactly_one=True`, and the corresponding deprecation warning is unnecessary.

File: geopy/geocoders/banfrance.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+        url = "?".join((self.reverse_api, urlencode(params)))
+        logger.debug("%s.reverse: %s", self.__class__.__name__, url)
+        return self._parse_json(
+            self._call_geocoder(url, timeout=timeout), exactly_one
+        )
+
+    @staticmethod
+    def parse_code(feature):
+        # TODO make this a private API
```
Comment (KostyaEsmukov) [inline geopy/geocoders/banfrance.py]: This note is not needed. And as it states, the parsing method name should be prefixed by an underscore, i.e. `_parse_location`

File: test/geocoders/banfrance.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+                "exactly_one": True},
+            {},
+        )
+        self.assertEqual(
+            res.address,
+            '15 Rue des Fontaines 89100 Collemiers'
+        )
+
+
+class BANFranceTestCase(BaseBANFranceTestCase, GeocoderTestBase):
```
Comment (KostyaEsmukov) [inline test/geocoders/banfrance.py]: There's no need in an abstract class for this geocoder, as it is one of its kind (unlike Nominatim). Could you please use a single class, like, for example, [test/geocoders/yandex.py](https://github.com/geopy/geopy/blob/3b19c8e4f6fdf13dddfbe7a3703ba031ec95194d/test/geocoders/yandex.py)?

File: test/geocoders/banfrance.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    def test_geocode_with_address(self):
+        """
+        banfrance.geocode Address
+        """
+        self.geocode_run(
+            {"query": "Camp des Landes, 41200 VILLEFRANCHE-SUR-CHER"},
+            {"latitude": 47.293048,
+                "longitude": 1.718985,
+                "address": "Le Camp des Landes 41200 Villefranche-sur-Cher"},
```
Comment (KostyaEsmukov) [inline test/geocoders/banfrance.py]: Please keep the dicts properly formatted.

**Comment by sheub** [comment]
Hi @KostyaEsmukov,
Thanks for the review, I did my best to make the corrections on time. Hope it's better now,
cheers,

Sheub

**Pull request merged**